CDK: Overview, getting started, concepts (#504)
0xPolygon/polygon-docs: docs/cdk/getting-started/index.md

> # Get Started

Reviewer: ```suggestion
# Get started
```

> This getting started guide will walk you through the process of setting up a Layer 2 blockchain using the Polygon CDK on your local machine, running the components in Docker containers.

Reviewer: ```suggestion
This _getting started_ guide walks you through the process of setting up a Layer 2 blockchain using the Polygon CDK on your local machine, running the components in Docker containers.
```

## Setting Up Your Environment

To run the Polygon CDK locally, the following prerequisites are required:

> **Hardware Requirements:**

Reviewer: ```suggestion
Hardware requirements:
```

- A Linux-based Operating System (or [WSL](https://learn.microsoft.com/en-us/windows/wsl/about))
- Minimum 8GB RAM and 2-core CPU
- An AMD64 architecture system

> **Software Dependencies:**

Reviewer: ```suggestion
Software Dependencies:
```

- [Docker Engine](https://docs.docker.com/engine/) (version 4.27 or higher)
- [Kurtosis CLI](https://docs.kurtosis.com/install/)
- [Foundry](https://book.getfoundry.sh/getting-started/installation)
- Optional: [yq](https://github.com/mikefarah/yq), [jq](https://stedolan.github.io/jq/), and [polycli](https://github.com/maticnetwork/polygon-cli) help submit transactions and interact with the environment.

> ## Exploring the CDK Kurtosis Package

Reviewer: ```suggestion
## Exploring the CDK kurtosis package
```

> The [Polygon CDK Kurtosis Package](https://github.com/0xPolygon/kurtosis-cdk/) allows you to easily customize and instantiate all of the components of a CDK chain. It uses the [Kurtosis](https://docs.kurtosis.com/) tool to orchestrate the setup of the chain components in Docker containers, with logic defined in [Starlark](https://github.com/bazelbuild/starlark) (a Python dialect) scripts to define the step-by-step process of setting up the chain.

Reviewer: ```suggestion
The [Polygon CDK kurtosis package](https://github.com/0xPolygon/kurtosis-cdk/) allows you to easily customize and instantiate all of the components of a CDK chain. It uses the [Kurtosis](https://docs.kurtosis.com/) tool to orchestrate the setup of the chain components in Docker containers, with logic defined in [Starlark](https://github.com/bazelbuild/starlark) (a Python dialect) scripts to define the step-by-step process of setting up the chain.
```

> ### Cloning the Repository

Reviewer: ```suggestion
### Cloning the repository
```

To get started, clone the repository and navigate to the `kurtosis-cdk` directory:

```bash
git clone https://github.com/0xPolygon/kurtosis-cdk.git
cd kurtosis-cdk
```

> ### Checking Your Environment

Reviewer: ```suggestion
### Checking your environment
```

> Ensure Docker is running on your machine, then run the following command to confirm all prerequisites are installed:

Reviewer: ```suggestion
Ensure Docker is running on your machine, then run the following command to confirm that all prerequisites are installed:
```

```bash
sh scripts/tool_check.sh
```

If everything is installed correctly, you should see the following output:

```bash
Checking that you have the necessary tools to deploy the Kurtosis CDK package...
✅ kurtosis is installed, meets the requirement (=0.89).
✅ docker is installed, meets the requirement (>=24.7).

You might as well need the following tools to interact with the environment...
✅ jq is installed.
✅ yq is installed, meets the requirement (>=3.2).
✅ cast is installed.
✅ polycli is installed.

🎉 You are ready to go!
```

> ### Customizing Your Chain

Reviewer: ```suggestion
### Customizing your chain
```

| 8           | Deploy the block explorer                                                                  | [blockscout.star](https://github.com/0xPolygon/kurtosis-cdk/blob/main/blockscout.star)                               | False              |
| 9           | Apply a load test to the chain                                                             | [load_test.star](https://github.com/0xPolygon/kurtosis-cdk/blob/main/workload.star)                                  | False              |
| 10          | Deploy [Blutgang](https://github.com/rainshowerLabs/blutgang) for load balancing & caching | [cdk_blutgang.star](https://github.com/0xPolygon/kurtosis-cdk/blob/main/cdk_blutgang.star)                           | False              |

You can customize (or skip) the logic for each of these steps by modifying the logic in the respective files.

#### params.yml

The `params.yml` file defines the parameters of the chain and the deployment process. It includes configurations for simple parameters such as the chain ID and more complex configurations such as the gas token smart contract address.

You can modify each of these parameters to customize the chain to your specific needs.

> ## Running the Chain Locally

Reviewer: ```suggestion
## Running the chain locally
```

     - `--args-file params.yml` specifies the configuration file to use for the deployment process.
     - `--image-download always` specifies to always download the latest Docker images for the deployment process.

This command typically takes around 10 minutes to complete and outputs the logs of each step in the deployment process for you to monitor the progress of the chain setup. Once the command is complete, you should see the following output:

```bash
Starlark code successfully run. No output was returned.

===============================================
||          Created enclave: cdk-v1          ||
===============================================
Name:            cdk-v1
Status:          RUNNING

========================================= Files Artifacts =========================================

... List of files generated during the deployment process ...

========================================== User Services ==========================================

... List of services with "RUNNING" status - none should be "FAILED"! ...

```

Run `kurtosis enclave inspect cdk-v1` to see the status of the enclave and the services running within it at any time.

> ## Interacting with the Chain

Reviewer: ```suggestion
## Interacting with the chain
```

Now that your chain is running, you can explore and interact with each component!

Below are a few examples of how you can interact with the chain.

> ### Sending Test Transactions

Reviewer: ```suggestion
### Sending test transactions
```

```bash
export ETH_RPC_URL="$(kurtosis port print cdk-v1 zkevm-node-rpc-001 http-rpc)"
```

Then, use `cast` to view information about the chain, such as the latest block number:

```bash
cast block-number
```

View the balance of an address, such as the pre-funded admin account:

```bash
cast balance --ether 0xE34aaF64b29273B7D567FCFc40544c014EEe9970
```

Send simple transactions to the chain, such as a transfer of some ETH:

```bash
cast send --legacy --value 0.01ether 0x0000000000000000000000000000000000000000 --private-key "0x12d7de8621a77640c9241b2595ba78ce443d05e94090365ab3bb5e19df82c625"
```

!!! info

      The `0xE34...9970` and `0x12d...c625` public-private key pair used in the above commands is the default admin account configured in `params.yml`.

### Load Testing the Chain

Use the [polycli loadtest](https://github.com/maticnetwork/polygon-cli/blob/main/doc/polycli_loadtest.md) command to send multiple transactions at once to the chain to test its performance:

```bash
polycli loadtest --rpc-url "$ETH_RPC_URL" --legacy --verbosity 700 --requests 500 --rate-limit 5 --mode t --private-key "0x12d7de8621a77640c9241b2595ba78ce443d05e94090365ab3bb5e19df82c625"
```

> ### Viewing Transaction Finality

Reviewer: ```suggestion
### Viewing transaction finality
```

A common way to check the status of the system is to ensure that batches are being sent and verified on the L1 chain.

Use `cast` to view the progression of batches from trusted, virtual, and verified states:

```bash
cast rpc zkevm_batchNumber          # Latest batch number on the L2
cast rpc zkevm_virtualBatchNumber   # Latest batch received on the L1
cast rpc zkevm_verifiedBatchNumber  # Latest batch verified or "proven" on the L1
```

> ### Opening the Bridge UI

Reviewer: ```suggestion
### Opening the bridge UI
```

To open the bridge interface and bridge tokens across the L1 and L2, run the following command:

```bash
open $(kurtosis port print cdk-v1 zkevm-bridge-proxy-001 bridge-interface)
```

> ### Viewing Chain Metrics

Reviewer: ```suggestion
### Viewing chain metrics
```

To view information such as how many transactions are being processed, the amount of gas being used, the time since a batch was last verified, how many addresses have bridged, and much more, a Grafana dashboard is included in the deployed observability stack which can be opened by running the following command:

```bash
open $(kurtosis port print cdk-v1 grafana-001 dashboards)
```

From the hamburger menu, navigate to `Dashboards` and select the `Panoptichain` dashboard to view all of the metrics.

![Panoptichain Dashboard](../../img/cdk/grafana.png)

> ### Stopping the Chain

Reviewer: ```suggestion
### Stopping the chain
```

If you want to **stop** the chain and remove all the containers, run the following command:

```bash
kurtosis clean --all
```

> ## Going to Production

Reviewer: ```suggestion
## Going to production
```
